samples/net/ptp/slave/src/ptp_slave.c

```diff
@@ -0,0 +1,313 @@
+		 * t4 + t2_t3_delay_ns / 2 + delay_ns
+		 */
+		int64_t current_ns = ptp_ts_net_to_ns(&state.t4) + t2_t3_delay_ns / 2 + delay_ns;
+		struct net_ptp_time current_t;
+
+		current_t.second = current_ns / NSEC_PER_SEC;
+		current_t.nanosecond = current_ns % NSEC_PER_SEC;
+
+		/* time delta is too large - jump */
+		ret = ptp_clock_set(state.clk, &current_t);
+		if (ret < 0) {
+			LOG_WRN("ptp_clock_set() failed: %d", ret);
+			return;
+		}
+		LOG_INF("PTP clock set: %lu.%09u", (unsigned long)current_t.second,
+			current_t.nanosecond);
+		state.last_offset_ns = 0;
+		return;
+	}
+
+	/* compensate the offset using a PD controller */
+	double rate = 1.0 + state.kp * (double)offset_ns / NSEC_PER_SEC +
+		      state.kd * (double)(offset_ns - state.last_offset_ns) / NSEC_PER_SEC;
+	ret = ptp_clock_rate_adjust(state.clk, rate);
+	if (ret < 0) {
+		LOG_WRN("ptp_clock_rate_adjust() failed: %d", ret);
+		return;
+	}
+	state.last_offset_ns = offset_ns;
+}
+
+static void pkt_received(struct net_context *context, struct net_pkt *pkt,
+			 union net_ip_header *ip_hdr, union net_proto_header *proto_hdr,
+			 int status, void *user_data)
+{
+	if (ntohs(NET_ETH_HDR(pkt)->type) != NET_ETH_PTYPE_PTP) {
+		goto done;
+	}
+
+	/* skip ethernet header */
+	net_pkt_skip(pkt, sizeof(struct net_eth_hdr));
+
+	/* read PTP header */
+	struct ptp_header ptp_hdr;
+
+	net_pkt_read(pkt, &ptp_hdr, sizeof(ptp_hdr));
+
+	if (ptp_hdr.version_ptp != 2 || ptp_hdr.minor_version_ptp != 0) {
+		LOG_WRN("Unsupposed PTP v%d.%d packet ignored", ptp_hdr.version_ptp,
+			ptp_hdr.minor_version_ptp);
+		goto done;
+	}
+
+	switch (ptp_hdr.message_type) {
+	case PTP_MESSAGE_TYPE_SYNC:
+		handle_ptp_sync(pkt, &ptp_hdr);
+		break;
+	case PTP_MESSAGE_TYPE_FOLLOW_UP:
+		handle_ptp_follow_up(pkt, &ptp_hdr);
+		break;
+	case PTP_MESSAGE_TYPE_DELAY_RESP:
+		handle_ptp_delay_resp(pkt, &ptp_hdr);
+		break;
+	}
+
+done:
+	net_pkt_unref(pkt);
+}
+
+int ptp_ts_get(struct net_ptp_time *net)
+{
+	return ptp_clock_get(state.clk, net);
+}
+
+int ptp_slave_init(struct net_if *iface)
+{
+	int ret;
+
+	memset(&state, 0, sizeof(state));
+
+	state.iface = iface;
+
+	state.clk = net_eth_get_ptp_clock(iface);
+	ret = device_is_ready(state.clk);
+	if (ret < 0) {
+		return ret;
+	}
+
+	struct net_linkaddr *iface_addr = net_if_get_link_addr(state.iface);
+
+	memcpy(state.port_identity, iface_addr->addr, iface_addr->len);
+
+	/* pid parameters determined experimentally */
+	state.kp = 0.3;
+	state.kd = 0.7;
+
+	ret = net_context_get(AF_PACKET, SOCK_RAW, ETH_P_ALL, &state.context);
+	if (ret < 0) {
+		return ret;
+	}
+
+	struct sockaddr_ll dst;
+
+	dst.sll_ifindex = net_if_get_by_iface(iface);
+	dst.sll_family = AF_PACKET;
+	dst.sll_protocol = htons(NET_ETH_PTYPE_PTP);
+
+	ret = net_context_bind(state.context, (const struct sockaddr *)&dst, sizeof(dst));
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = net_context_recv(state.context, pkt_received, K_NO_WAIT, NULL);
+	if (ret < 0) {
+		return ret;
+	}
```

Comment:
We try to avoid using `net_context` APIs in the samples as the BSD socket interface is meant for application use and can be used here.

Reply:
This code requires receiving all PTP L2 ethernet packets, does Zephyr's BSD socket API support that use case? Is there an example?

Reply:
For raw socket support, see `samples/net/sockets/packet` for an example.

Reply:
Thanks for the pointer. I believe last time I tried this, PTP packets were confused with gPTP and dropped - so this might require some more modifications to ethernet stack. We can explore that as part of adding this example in a follow-on PR.

Reply:
Ah, it might be that gPTP stack is stealing the packets but that should have happened when using `net_context` API too. Internally BSD socket code is just calling net_context API.

Reply:
I've started playing with that. Question - how to accomplish this with sockets API? `net_pkt_timestamp(pkt)`